README.md
# SiriDB HTTP
![alt SiriDB HTTP](/siridb-http.png?raw=true)

SiriDB HTTP provides a HTTP API and optional web interface for [SiriDB](https://github.com/SiriDB/siridb-server).

> Note: Since version 2.0.0 SiriDB HTTP is written in Go. For the 1.x version in Python you should go to
> this [release tag](https://github.com/SiriDB/siridb-http/tree/1.0.7).

---------------------------------------
  * [Features](#features)
  * [Installation](#installation)
    * [Pre-compiled](#pre-compiled)
    * [Compile from source](#compile-from-source)
  * [Configuration](#configuration)
    * [Autorun on startup](#autorun-on-startup)
    * [Multi server support](#multi-server-support)
  * [HTTP API](#http-api)
    * [Content Types](#content-types)
    * [Database info](#database-info)
    * [Authentication](#authentication)
      * [Session authentication](#session-authentication)
        * [Fetch](#fetch)
        * [Login](#login)
        * [Logout](#logout)
      * [Basic authentication](#basic-authentication)
    * [Query](#query)
    * [Insert](#insert)
      * [JSON, MsgPack, QPack](#insert-json)
      * [CSV](#insert-csv)
        * [List format](#list-format)
        * [Table format](#table-format)
  * [Socket.io](#socketio)
  * [Web Interface](#web-interface)
  * [SSL (HTTPS)](#ssl-https)
---------------------------------------

## Features
  - Optional Web interface for sending queries and inserting data
  - SSL (HTTPS) support
  - Optional Basic Authentication
  - Support for JSON, MsgPack, QPack and CSV
  - IPv6 support
  - Socket.io support


## Installation
SiriDB HTTP 2.x can be compiled from source or, for most systems, you can simply download a pre-compiled binary.

### Pre-compiled
Go to https://github.com/SiriDB/siridb-http/releases/latest and download the binary for your system.
In this documentation we refer to the binary as `siridb-http`. On Linux/OSX it might be required to set the execution flag:
```
$ chmod +x siridb-http_X.Y.Z_OS_ARCH.bin
```

You might want to copy the binary to /usr/local/bin and create a symlink like this:
```
$ sudo cp siridb-http_X.Y.Z_OS_ARCH.bin /usr/local/bin/
$ sudo ln -s /usr/local/bin/siridb-http_X.Y.Z_OS_ARCH.bin /usr/local/bin/siridb-http
```
> Note: replace `X.Y.Z_OS_ARCH` with your binary, for example `2.0.0_linux_amd64`

### Compile from source
> Before compiling from source make sure **go**, **npm** and **git** are installed and your [$GOPATH](https://github.com/golang/go/wiki/GOPATH) is set.

Clone the project using git. (we assume git is installed)
```
git clone https://github.com/SiriDB/siridb-http
```

Make sure less is installed:
```
$ sudo npm install -g less less-plugin-clean-css
```

The gobuild.py script can be used to build the binary:
```
$ ./gobuild.py -i -l -w -b -p
```

Or, if you want the development version which uses original files from /build and /static instead of build-in files:
```
$ ./gobuild.py -i -l -w -b -d
```

## Configuration
For running SiriDB HTTP a configuration file is required and should be provided using the `-c` or `--config` argument. The easiest way to create a configuration file is to save the output from
siridb-http to a file:

> Note: you might want to switch to root and later create a service to automatically start SiriDB HTTP at startup.

Switch to root or skip this step if you want to save the configuration file with your current user.
```
$ sudo su -
```

Save a template configuration file to for example ~/.siridb-http.conf.
```
$ siridb-http > ~/.siridb-http.conf
```

Now edit the file with you favorite editor and at least set the `user`, `password` and `dbname`.
When the configuration is saved you can start the server using:
```
$ siridb-http -c ~/.siridb-http.conf
```

### Autorun on startup
Depending on your OS and subsystem you can create a service to start SiriDB HTTP.
This is an example of how to do this using systemd which is currently the default for Ubuntu:

First create the service file: (you might need to change the ExecStart line)
```
$ sudo cat > /lib/systemd/system/siridb-http.service <<EOL
[Unit]
Description=SiriDB HTTP Service
After=network.target

[Service]
ExecStart=/usr/local/bin/siridb-http --config ~/.siridb-http.conf
StandardOutput=journal
TimeoutStartSec=3
TimeoutStopSec=5

[Install]
WantedBy=multi-user.target
EOL
```

Now reload the daemon and start the service
```
sudo systemctl daemon-reload
sudo systemctl start siridb-http.service
```

### Multi server support
SiriDB can scale across multiple pools and can be made high-available by adding two servers to each pool. For example you could have four siridb servers sdb01, sdb02, sdb03 and sdb04 all listening to port 9000. In this example we assume sdb01 and sdb02 are members of `pool 0` and sdb03 and sdb04 are members of `pool 1`.

We should now configure SiriDB to connect to both servers in pool 0 and/or pool 1. This ensures that queries and inserts always work, even when a server in the SiriDB cluster is not available for whatever reason. The only requirement is that each pool has at least one server online.

To configure SiriDB HTTP to connect to multiple servers a comma must be used as separator like this:
```
[Database]
...
servers = sdb01,sdb02,sdb03,sdb04
```

## API
SiriDB HTTP provides both a REST API and Socket.io support which can be used to insert and query a SiriDB cluster.

### Content Types
The following URIs have support for JSON, MsgPack, QPack and CSV:
- /db-info
- /auth/fetch
- /auth/login
- /auth/logout
- /query
- /insert

In most examples below we use JSON but this format is fully compatible with MsgPack and QPack so it
should be easy to translate. CSV on the other hand is different and unless explained, each request and
response can be transfomed to a key,value per line. See the example below:
```json
{
    "username": "my_username",
    "password": "my_password"
}
```

The above translates to the following CSV:
```
username,my_username
password,my_password
```

> Note: when a string in CSV contains a comma (,) then the string must be wrapped between double quotes.
> If double quotes are also required in a string, the double quote should be escaped with a second double quote.

### Database info
With the `/db-info` URI, database and version information can be asked.
```
type:      GET or POST
uri:       /db-info
header:    Content-Type:  'application/json'
```

Response:
```json
{
    "dbname": "my_database-name",
    "timePrecision": "database time precision: s, ms, us or ns",
    "version": "SiriDB Server version, for example: 2.0.18",
    "httpServer": "SiriDB HTTP version, for example: 2.0.0"
}
```

> Note that `version` does not guarantee that each SiriDB server in a cluster is running the same version.

### Authentication
Authentication is required when `require_authentication` is set to `True` in the configuration file. When authentication is not required, the `/insert` and `/query` URIs can be used directly without any authentication as long as the user configured in the configuration file has privileges to perform the request.

#### Session authentication
SiriDB HTTP has session support and exposes the following URIs for handling session authentication:
- /auth/fetch
- /auth/login
- /auth/logout

> Note: in the examples below we use 'application/json' as Content-Type but the following alternatives
> are also allowed: 'application/x-msgpack', 'application/x-qpack' and 'application/csv'.

##### Fetch
Fetch can be used to retrieve the current session user.
```
type:      GET or POST
uri:       /auth/fetch
header:    Content-Type:  'application/json'
```
The response contains the current user and a boolean value to indicate if authentication is required. In case no user is logged on and authentication is required, the value for `user` will be `null`.

Example response:
```json
{
    "user": "logged_on_username_or_null",
    "authRequired": true
}
```

##### Login
Login can be used to authenticate and create a SiriDB database user. If the option `enable_multi_user` in the configuration file is set to `True`, any database user can be used. In case multi user support is turned off, the only allowed user is the one configured in the configuration file.

```
type:      POST
uri:       /auth/login
header:    Content-Type:  'application/json'
body:      {"username": <my-username>, "password": <my-secret-password>}
```

Success response:
```json
{
    "user": "logged_on_username"
}
```

In case authentication has failed, error code 422 will be returned and the body will contain an appropriate error message.

##### Logout
When calling this uri the current session will be cleared.
```
type:      GET or POST
uri:       /auth/logout
header:    Content-Type:  'application/json'
```
Response:
```json
{
    "user": null
}
```

#### Basic authentication
As an alternative to session authentication it is possible to use basic authentication. To allow basic authentication the option `enable_basic_auth` must be set to `True` in the configuration file.

Example Authorization header for username *iris* with password *siri*:
```
Authorization: Basic aXJpczpzaXJp
```

### Query
The `/query` POST handler can be used for querying SiriDB. SiriDB HTTP supports multiple formats that can be used by setting the `Content-Type` in the header.
```
type:      POST
uri:       /query
header:    Content-Type:  'application/json'
body:      {"query": <query string>, "timeout": <optional timout in seconds>}
```

Example body:
```json
{
    "query": "select mean(1h) => difference() from 'my-series'"
}
```

### Insert
The `/insert` POST handler can be used for inserting data into SiriDB. The same content types as for queries are supported. Both MsgPack and QPack are similar to JSON except that the data is packed to a byte string. Therefore we only explain JSON and CSV data here. *(Note: in the examples below we use a second time-precision)*

#### Insert JSON
The preferred json layout is as following: (this is the layout which is returned by SiriDB on a select query)
```json
{
    "my-series-01": [[1493126582, 4.2], ...],
    ...
}
```

Optionally the following format can be used:
```json
[
    {
        "name": "my-series-01",
        "points": [[1493126582, 4.2], ...]
    },
    ...
]
```

#### Insert CSV
CSV data is allowed in two formats which we call the list and table format.

##### List format
When using the list format, each row in the csv should contain a series name, timestamp and value.

Example list:
```csv
Series 001,1440138931,100
Series 003,1440138931,8.0
Series 001,1440140932,40
Series 002,1440140932,9.3
```
##### Table format
A table format is more compact, especially if multiple series share points with equal timestamps. The csv should start with an empty field that is indicated with the first comma.

Example table:
```csv
,Series 001,Series 002,Series 003
1440138931,100,,8.0
1440140932,40,9.3,
```

## Socket.io
SiriDB HTTP has socket.io support when `enable_socket_io` is set to `True` in the configuration file.

For Socket.io the following events are available:
- `db-info`
- `auth fetch`
- `auth login`
- `auth logout`
- `query`
- `insert`

The result for an event contains a status code and data object. The status codes are equal to HTTP status codes. For example the success code is 200.
When the status code is anything other than 200, the data object will be a string representing the error message.

Example of using Socket.io with HTML/JavaScript:
```html
<html>
<head>
<script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.1/socket.io.js"></script>
</head>
<body>
<script>
// create socket, assuming SiriDB HTTP is running on localhost and listening to port 5050
var socket = io.connect("http://127.0.0.1:5050");

// get database and version information
socket.emit('db-info', null, function (status, data) {
    console.log(status, data);
});

// authenticate using user iris with password siri
socket.emit('auth login', {username: "iris", password: "siri"}, function (status, data) {
    console.log(status, data);

    // if successful perform a query
    if (status == 200) {
        socket.emit('query', {query: "list series limit 5"}, function (status, data) {
            console.log(status, data);
        });
    }
});
</script>
</body>
</html>
```

## Web interface
SiriDB has an optional web interface that can be enabled by setting `enable_web` to `True`. This web interface will ask for user authentication if `enable_authentication` is set to `True`. Only the `user` that is configured in the configuration file is allowed to login unless `enable_multi_user` is set to `True`.

The Web interface allows you to run queries and insert data using JSON format.

## SSL (HTTPS)
SSL (HTTPS) support can be enabled by setting `enable_ssl` to `True`. When enabled the `crt_file` and `key_file` in section `[SSL]` must be set.



ChangeLog
siridb-http (2.0.8)

  * Update SiriDB grammar to version 2.0.29

  * Set document title to database name

  * Replace Reflux with Vlow

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  09 Jul 2018

siridb-http (2.0.7)

  * Update SiriDB grammar to version 2.0.28

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  20 Jun 2018

siridb-http (2.0.6)

  * Update SiriDB grammar to version 2.0.27

  * Added log (string) support.

  * Update npm packages and removed obsolete packages.

  * Added option to switch from charts to tables and back.

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  24 May 2018

siridb-http (2.0.5)

  * Replace parseInt with Math.floor()

  * Upgrade some npm packages

  * Re-factored React PropTypes

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  06 Mar 2018

siridb-http (2.0.4)

  * Update to Go SiriDB Connector

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  09 Nov 2017

siridb-http (2.0.3)

  * Update grammar with new properties (issue #16)

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  15 Aug 2017

siridb-http (2.0.2)

  * Changed text when serving from 127.0.0.1 to 0.0.0.0.

  * Fixed buffer bug in go-siridb-connector. (issue #14)

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  20 Jul 2017

siridb-http (2.0.1)

  * Implemented go siridb connector 1.0.5 which fixed a bug for large packages.

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  02 Jun 2017

siridb-http (2.0.0)

  * Initial version created with the Go language.

  * See archive for changes to the 1.x Python versions.

 -- Jeroen van der Heijden <jeroen@transceptor.technology>  31 May 2017



handlers.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"reflect"
	"strconv"
	"strings"

	qpack "github.com/transceptor-technology/go-qpack"
	siridb "github.com/transceptor-technology/go-siridb-connector"
	"github.com/transceptor-technology/go-socket.io"
	msgpack "gopkg.in/vmihailenco/msgpack.v2"
)

// StatusUnprocessableEntity is not available in Go 1.6 and older
const StatusUnprocessableEntity int = 422

type tDb struct {
	Dbname        string `json:"dbname" qp:"dbname" msgpack:"dbname" csv:"dbname"`
	TimePrecision string `json:"timePrecision" qp:"timePrecision" msgpack:"timePrecision" csv:"timePrecision"`
	Version       string `json:"version" qp:"version" msgpack:"version" csv:"version"`
	HTTPServer    string `json:"httpServer" qp:"httpServer" msgpack:"httpServer" csv:"httpServer"`
}

type tAuthFetch struct {
	User         interface{} `json:"user" qp:"user" msgpack:"user" csv:"user"`
	AuthRequired bool        `json:"authRequired" qp:"authRequired" msgpack:"authRequired" csv:"authRequired"`
}

type tAuthLoginReq struct {
	Username string `json:"username" qp:"username" msgpack:"username" csv:"username"`
	Password string `json:"password" qp:"password" msgpack:"password" csv:"password"`
}

type tAuthLoginRes struct {
	User string `json:"user" qp:"user" msgpack:"user" msgpack:"user" csv:"user"`
}

type tAuthLogoff struct {
	User interface{} `json:"user" qp:"user" msgpack:"user" csv:"user"`
}

type tQuery struct {
	Query   string      `json:"query" qp:"query" msgpack:"query" csv:"query"`
	Timeout interface{} `json:"timeout" qp:"timeout" msgpack:"timeout" csv:"timeout"`
}

func checkBasicAuth(r *http.Request) (conn *Conn) {
	if !base.enableBasicAuth {
		return nil
	}
	username, password, ok := r.BasicAuth()
	if !ok {
		return nil
	}
	conn = getConnByUser(username)
	if conn != nil {
		if password == conn.password {
			return conn
		}
		return nil
	}
	if base.multiUser {
		conn, _ = addConnection(username, password)
	}
	return conn
}

func getConnBySIO(so *socketio.Socket) (conn *Conn, err error) {
	if user, ok := base.ssessions[(*so).Id()]; ok {
		conn = getConnByUser(user)
		if conn == nil {
			err = fmt.Errorf("no connection for user '%s' found, please try to login again", user)
		}
	} else if base.reqAuth {
		err = fmt.Errorf("not authenticated")
	} else {
		conn = &base.connections[0]
	}
	return conn, err
}

func getConnByHTTP(w http.ResponseWriter, r *http.Request) *Conn {
	var conn *Conn

	if conn = checkBasicAuth(r); conn != nil {
		return conn
	}
	sess, err := base.gsessions.SessionStart(w, r)

	if err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
	} else {
		user, ok := sess.Get("user").(string)
		if ok {
			conn = getConnByUser(user)
			if conn == nil {
				sendError(
					w,
					fmt.Sprintf("no connection for user '%s' found, please try to login again", user),
					http.StatusUnauthorized)
			}
		} else if base.reqAuth {
			sendError(w, "not authenticated", http.StatusUnauthorized)
		} else {
			conn = &base.connections[0]
		}
	}
	return conn
}

func sendError(w http.ResponseWriter, err string, code int) {
	w.Header().Set("Access-Control-Allow-Origin", "*")
	http.Error(w, err, code)
}

func sendCSV(w http.ResponseWriter, data interface{}) {
	if s, err := toCsv(data); err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
	} else {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Content-Type", "application/csv; charset=UTF-8")
		w.Write([]byte(s))
	}
}

func sendJSON(w http.ResponseWriter, data interface{}) {
	if b, err := json.Marshal(data); err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
	} else {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Content-Type", "application/json; charset=UTF-8")
		w.Write(b)
	}
}

func sendQPack(w http.ResponseWriter, data interface{}) {
	if b, err := qpack.Pack(data); err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
	} else {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Content-Type", "application/x-qpack; charset=UTF-8")
		w.Write(b)
	}
}

func sendMsgPack(w http.ResponseWriter, data interface{}) {
	if b, err := msgpack.Marshal(data); err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
	} else {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Content-Type", "application/x-msgpack; charset=UTF-8")
		w.Write(b)
	}
}

func getConnByUser(user string) *Conn {
	for _, conn := range base.connections {
		if conn.user == user {
			return &conn
		}
	}
	return nil
}

func addConnection(username, password string) (*Conn, error) {
	var conn Conn
	conn.user = username
	conn.password = password
	conn.client = siridb.NewClient(
		conn.user,                        // user
		conn.password,                    // password
		base.dbname,                      // database
		serversToInterface(base.servers), // siridb server(s)
		base.logCh,                       // optional log channel
	)
	conn.client.Connect()
	if conn.client.IsConnected() && conn.client.IsAvailable() {
		base.connections = append(base.connections, conn)
	} else {
		conn.client.Close()
		return nil, fmt.Errorf("Cannot login using the username and password")
	}
	return &conn, nil
}

func handlerNotFound(w http.ResponseWriter, r *http.Request) {
	sendError(w, "404 not found", http.StatusNotFound)
}

func onDbInfo(so *socketio.Socket) (int, interface{}) {
	db := tDb{
		Dbname:        base.dbname,
		TimePrecision: base.timePrecision,
		Version:       base.version,
		HTTPServer:    AppVersion}
	return http.StatusOK, db
}

func handlerDbInfo(w http.ResponseWriter, r *http.Request) {
	db := tDb{
		Dbname:        base.dbname,
		TimePrecision: base.timePrecision,
		Version:       base.version,
		HTTPServer:    AppVersion}
	sendData(w, r, db)
}

func sendData(w http.ResponseWriter, r *http.Request, data interface{}) {
	contentType := r.Header.Get("Content-type")

	switch strings.ToLower(contentType) {
	case "application/csv":
		sendCSV(w, data)
	case "application/json":
		sendJSON(w, data)
	case "application/x-qpack":
		sendQPack(w, data)
	case "application/x-msgpack":
		sendMsgPack(w, data)
	default:
		sendError(w, fmt.Sprintf("unsupported content-type: %s", contentType), http.StatusUnsupportedMediaType)
	}
}

func readBody(w http.ResponseWriter, r *http.Request, v interface{}) error {
	contentType := r.Header.Get("Content-type")

	switch strings.ToLower(contentType) {
	case "application/csv":
		return readCSV(w, r, &v)
	case "application/json":
		return readJSON(w, r, &v)
	case "application/x-qpack":
		return readQPack(w, r, &v)
	case "application/x-msgpack":
		return readMsgPack(w, r, &v)
	default:
		err := fmt.Errorf("unsupported content-type: %s", contentType)
		sendError(w, err.Error(), http.StatusUnsupportedMediaType)
		return err
	}
}

func onAuthFetch(so *socketio.Socket) (int, interface{}) {
	authFetch := tAuthFetch{User: nil, AuthRequired: base.reqAuth}

	if user, ok := base.ssessions[(*so).Id()]; ok && getConnByUser(user) != nil {
		authFetch.User = user
	} else if !base.reqAuth {
		authFetch.User = base.connections[0].user
	}
	return http.StatusOK, authFetch
}

func handlerAuthFetch(w http.ResponseWriter, r *http.Request) {

	authFetch := tAuthFetch{User: nil, AuthRequired: base.reqAuth}
	sess, err := base.gsessions.SessionStart(w, r)

	if err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return
	}

	if user, ok := sess.Get("user").(string); ok && getConnByUser(user) != nil {
		authFetch.User = user
	} else if !base.reqAuth {
		authFetch.User = base.connections[0].user
	}
	sendData(w, r, authFetch)
}

func onAuthLogin(so *socketio.Socket, req *tAuthLoginReq) (int, interface{}) {
	if conn := getConnByUser(req.Username); conn != nil {
		if req.Password != conn.password {
			return StatusUnprocessableEntity, "Username or password incorrect"
		}
	} else if base.multiUser {
		if _, err := addConnection(req.Username, req.Password); err != nil {
			return StatusUnprocessableEntity, err.Error()
		}
	} else {
		return StatusUnprocessableEntity, "Multiple user login is not allowed"
	}

	base.ssessions[(*so).Id()] = req.Username
	authLoginRes := tAuthLoginRes{User: req.Username}

	return http.StatusOK, authLoginRes
}

func handlerAuthLogin(w http.ResponseWriter, r *http.Request) {

	var authLoginReq tAuthLoginReq

	sess, err := base.gsessions.SessionStart(w, r)
	if err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return
	}

	if err := readBody(w, r, &authLoginReq); err != nil {
		return // error is send by the readBody function
	}

	if conn := getConnByUser(authLoginReq.Username); conn != nil {
		if authLoginReq.Password != conn.password {
			sendError(w, "Username or password incorrect", StatusUnprocessableEntity)
			return
		}
	} else if base.multiUser {
		if _, err := addConnection(authLoginReq.Username, authLoginReq.Password); err != nil {
			sendError(w, err.Error(), StatusUnprocessableEntity)
			return
		}
	} else {
		sendError(w, "Multiple user login is not allowed", StatusUnprocessableEntity)
		return
	}

	sess.Set("user", authLoginReq.Username)
	authLoginRes := tAuthLoginRes{User: authLoginReq.Username}
	sendData(w, r, authLoginRes)
}

func onAuthLogout(so *socketio.Socket) (int, interface{}) {
	authLogoff := tAuthLogoff{User: nil}
	delete(base.ssessions, (*so).Id())
	return http.StatusOK, authLogoff
}

func handlerAuthLogout(w http.ResponseWriter, r *http.Request) {
	authLogoff := tAuthLogoff{User: nil}

	sess, err := base.gsessions.SessionStart(w, r)
	if err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return
	}

	if err = sess.Flush(); err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return
	}

	sendData(w, r, authLogoff)
}

func onQuery(so *socketio.Socket, req *tQuery) (int, interface{}) {
	conn, err := getConnBySIO(so)
	if err != nil {
		return http.StatusUnauthorized, err.Error()
	}

	var timeout uint64
	timeout = 30

	if req.Timeout != nil {
		if timeout, err = strconv.ParseUint(fmt.Sprint(req.Timeout), 10, 16); err != nil {
			timeout = 30
		}
	}

	res, err := conn.client.Query(req.Query, uint16(timeout))
	if err != nil {
		return http.StatusInternalServerError, err.Error()
	}

	return http.StatusOK, res
}

func handlerQuery(w http.ResponseWriter, r *http.Request) {
	if conn := getConnByHTTP(w, r); conn != nil {
		var query tQuery

		if err := readBody(w, r, &query); err != nil {
			return // error is send by the readBody function
		}

		timeout, ok := query.Timeout.(uint16)
		if !ok {
			timeout = 30
		}

		res, err := conn.client.Query(query.Query, timeout)
		if err != nil {
			sendError(w, err.Error(), http.StatusInternalServerError)
			return
		}

		sendData(w, r, res)
	}
}

func onInsert(so *socketio.Socket, insert *interface{}) (int, interface{}) {
	conn, err := getConnBySIO(so)
	if err != nil {
		return http.StatusUnauthorized, err.Error()
	}

	res, err := conn.client.Insert(*insert, base.insertTimeout)
	if err != nil {
		return http.StatusInternalServerError, err.Error()
	}

	return http.StatusOK, res
}

func readJSON(w http.ResponseWriter, r *http.Request, v *interface{}) error {
	decoder := json.NewDecoder(r.Body)
	decoder.UseNumber()
	if err := decoder.Decode(v); err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return err
	}
	return nil
}

func readMsgPack(w http.ResponseWriter, r *http.Request, v *interface{}) error {
	decoder := msgpack.NewDecoder(r.Body)
	if err := decoder.Decode(v); err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return err
	}
	return nil
}

func resToPlan(w http.ResponseWriter, res interface{}, v *interface{}, ft string) error {
	iface, ok := (*v).(*interface{})
	if ok {
		*iface = res
		return nil
	}

	m, ok := res.(map[string]interface{})
	if !ok {
		err := fmt.Errorf("expecting a map for a non interface{} value")
		sendError(w, err.Error(), http.StatusInternalServerError)
		return err
	}

	e := reflect.ValueOf(*v).Elem()
	t := e.Type()
	n := t.NumField()

	for i := 0; i < n; i++ {
		field := t.Field(i)
		fn := field.Tag.Get(ft)
		if len(fn) == 0 {
			fn = field.Name
		}
		val, ok := m[fn]
		if ok {
			if e.Field(i).Type() != reflect.TypeOf(val) {
				err := fmt.Errorf("unexpected type")
				sendError(w, err.Error(), http.StatusInternalServerError)
				return err
			}
			e.Field(i).Set(reflect.ValueOf(val))
		}
	}
	return nil
}

func readQPack(w http.ResponseWriter, r *http.Request, v *interface{}) error {
	b, err := ioutil.ReadAll(r.Body)

	if err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return err
	}

	res, err := qpack.Unpack(b, qpack.QpFlagStringKeysOnly)
	if err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return err
	}

	return resToPlan(w, res, v, "qp")
}

func readCSV(w http.ResponseWriter, r *http.Request, v *interface{}) error {
	res, err := parseCsv(r.Body)
	if err != nil {
		sendError(w, err.Error(), http.StatusInternalServerError)
		return err
	}
	return resToPlan(w, res, v, "csv")
}

func handlerInsert(w http.ResponseWriter, r *http.Request) {
	if conn := getConnByHTTP(w, r); conn != nil {
		var insert interface{}
		if err := readBody(w, r, &insert); err != nil {
			return // error is send by the readBody function
		}

		res, err := conn.client.Insert(insert, base.insertTimeout)
		if err != nil {
			sendError(w, err.Error(), http.StatusInternalServerError)
			return
		}

		sendData(w, r, res)
	}
}
